File: 2026/04-April/01-Wednesday/solution.py
from typing import List

class Solution:
    def survivedRobotsHealths(self, positions: List[int], healths: List[int], directions: str) -> List[int]:
        n = len(positions)
        robots = [(positions[i], healths[i], directions[i], i) for i in range(n)]
        
        robots.sort()
        stack = []

        for pos, health, d, idx in robots:
            if d == 'R':
                stack.append([pos, health, d, idx])
            else:
                while stack and stack[-1][2] == 'R':
                    if stack[-1][1] < health:
                        stack.pop()
                        health -= 1
                    elif stack[-1][1] > health:
                        stack[-1][1] -= 1
                        health = 0
                        break
                    else:
                        stack.pop()
                        health = 0
                        break
                if health > 0:
                    stack.append([pos, health, d, idx])

        stack.sort(key=lambda x: x[3])
        return [r[1] for r in stack]

File: 2026/04-April/01-Wednesday/solution.java
import java.util.*;

class Solution {
    public List<Integer> survivedRobotsHealths(int[] positions, int[] healths, String directions) {
        int n = positions.length;
        int[][] robots = new int[n][4];

        for (int i = 0; i < n; i++) {
            robots[i] = new int[]{positions[i], healths[i], directions.charAt(i) == 'R' ? 1 : 0, i};
        }

        Arrays.sort(robots, Comparator.comparingInt(a -> a[0]));

        Stack<int[]> stack = new Stack<>();

        for (int[] r : robots) {
            if (r[2] == 1) {
                stack.push(r);
            } else {
                while (!stack.isEmpty() && stack.peek()[2] == 1) {
                    int[] top = stack.peek();
                    if (top[1] < r[1]) {
                        stack.pop();
                        r[1]--;
                    } else if (top[1] > r[1]) {
                        top[1]--;
                        r[1] = 0;
                        break;
                    } else {
                        stack.pop();
                        r[1] = 0;
                        break;
                    }
                }
                if (r[1] > 0) stack.push(r);
            }
        }

        List<int[]> survivors = new ArrayList<>(stack);
        survivors.sort(Comparator.comparingInt(a -> a[3]));

        List<Integer> res = new ArrayList<>();
        for (int[] r : survivors) res.add(r[1]);

        return res;
    }
}
